Data/HashMap/Strict.hs

```diff
@@ -249,6 +254,54 @@ alter f k m =
     Just v  -> insert k v m
 {-# INLINABLE alter #-}
 
+-- | /O(log n)/  The expression (@'alterF' f k map@) alters the value @x@ at
+-- @k@, or absence thereof. @alterF@ can be used to insert, delete, or update
+-- a value in a map.
+--
+-- Note: 'alterF' is a flipped version of the 'at' combinator from
+-- 'Control.Lens.At'.
+--
+-- @since 0.2.9
+alterF :: (Functor f, Eq k, Hashable k)
+       => (Maybe v -> f (Maybe v)) -> k -> HashMap k v -> f (HashMap k v)
+-- Special care is taken to only calculate the hash and only perform a key
+-- comparison once.
+alterF f k m = (<$> f mv) $ \fres ->
+  case fres of
+
+    ------------------------------
+    -- Delete the key from the map.
+    Nothing -> case lookupRes of
+
+      -- Key did not exist in the map to begin with, no-op
+      Absent -> m
+
+      -- Key did exist, no collision
+      Present _ collPos -> deleteKeyExists collPos h k 0 m
+
+    ------------------------------
+    -- Update value
+    Just v' -> case lookupRes of
+
+      -- Key did not exist before, insert v' under a new key
+      Absent -> insertNewKey h k v' 0 m
+
+      -- Key existed before, no hash collision
+      Present v collPos ->
+        -- TODO(m-renaud): Verify ptrEq is valid here.
+        if v `ptrEq` v'
+        -- If the value is identical, no-op
+        then m
+        -- If the value changed, update the value.
+        else v' `seq` insertKeyExists collPos h k v' 0 m
```

Comment:
@treeowl Is the use of `seq` here fine or should I just use bang patterns on the `Just` pattern match above. My thought was it may be more efficient to wait until after the `ptrEq` check but it may not matter. WDYT?